File: app/src/main/java/com/example/mobile6/ui/medicine/MedicineSearchFragment.java
package com.example.mobile6.ui.medicine;

import android.os.Bundle;
import android.text.Editable;
import android.text.TextWatcher;
import android.view.LayoutInflater;
import android.view.ViewGroup;

import androidx.annotation.Nullable;
import androidx.recyclerview.widget.LinearLayoutManager;

import com.example.mobile6.MainApplication;
import com.example.mobile6.R;
import com.example.mobile6.adapter.MedicineAdapter;
import com.example.mobile6.dao.MedicineDao;
import com.example.mobile6.databinding.FragmentMedicineSearchBinding;
import com.example.mobile6.model.Medicine;
import com.example.mobile6.ui.base.BaseFragment;

import java.util.ArrayList;
import java.util.List;
import java.util.concurrent.ExecutorService;

public class MedicineSearchFragment extends BaseFragment<FragmentMedicineSearchBinding> {
    private MedicineAdapter medicineAdapter;
    private MedicineDao medicineDao;
    private ExecutorService diskIO;

    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        // Khởi tạo database
        medicineDao = ((MainApplication) requireActivity().getApplication()).getDatabase().medicineDao();
        diskIO = ((MainApplication) requireActivity().getApplication()).getAppExecutors().diskIO();
    }

    @Override
    protected FragmentMedicineSearchBinding inflateBinding(LayoutInflater layoutInflater, ViewGroup container) {
        return FragmentMedicineSearchBinding.inflate(layoutInflater, container, false);
    }

    @Override
    protected void initViews() {
        super.initViews();
        setupUI();
        loadDummyData();
    }

    private void setupUI() {
        // Setup RecyclerView
        medicineAdapter = new MedicineAdapter(medicine -> {
            Bundle bundle = new Bundle();
            bundle.putString("medicineId", medicine.getId());
            navigateTo(R.id.action_medicineSearchFragment_to_medicineDetailFragment, bundle);
        });
        binding.medicinesRecyclerView.setLayoutManager(new LinearLayoutManager(requireContext()));
        binding.medicinesRecyclerView.setAdapter(medicineAdapter);

        // Setup search functionality
        binding.searchEditText.addTextChangedListener(new TextWatcher() {
            @Override
            public void beforeTextChanged(CharSequence s, int start, int count, int after) {
                // Not needed
            }

            @Override
            public void onTextChanged(CharSequence s, int start, int before, int count) {
                // Tìm kiếm thuốc theo tên
                filterMedicines(s.toString());
            }

            @Override
            public void afterTextChanged(Editable s) {
                // Not needed
            }
        });

        // Setup back button click
        binding.backButton.setOnClickListener(v -> back());
    }

    private void loadDummyData() {
        try {
            // Tạo dữ liệu mẫu và lưu vào cơ sở dữ liệu
            List<Medicine> medicines = new ArrayList<>();
            medicines.add(Medicine.builder().id("1").name("Panadol Extra 500mg").ingredients("Paracetamol, Caffeine").build());
            medicines.add(Medicine.builder().id("2").name("Efferalgan 500mg").ingredients("Paracetamol").build());
            medicines.add(Medicine.builder().id("3").name("Amoxicillin 500mg").ingredients("Amoxicillin").build());

            // Hiển thị trước (không phụ thuộc database)
            medicineAdapter.submitList(medicines);

            // Lưu vào database
            diskIO.execute(() -> {
                try {
                    medicineDao.insertMedicines(medicines);
                } catch (Exception e) {
                    e.printStackTrace();
                }
            });
        } catch (Exception e) {
            e.printStackTrace();
        }
    }

    private void filterMedicines(String query) {
        try {
            if (query.isEmpty()) {
                // Nếu không có từ khóa, lấy tất cả
                diskIO.execute(() -> {
                    try {
                        List<Medicine> medicines = medicineDao.getAllMedicines();
                        if (getActivity() != null) {
                            getActivity().runOnUiThread(() -> medicineAdapter.submitList(medicines));
                        }
                    } catch (Exception e) {
                        e.printStackTrace();
                    }
                });
            } else {
                // Tìm kiếm theo từ khóa
                diskIO.execute(() -> {
                    try {
                        List<Medicine> filteredMedicines = medicineDao.searchMedicines(query);
                        if (getActivity() != null) {
                            getActivity().runOnUiThread(() -> medicineAdapter.submitList(filteredMedicines));
                        }
                    } catch (Exception e) {
                        e.printStackTrace();
                    }
                });
            }
        } catch (Exception e) {
            e.printStackTrace();
        }
    }
} 

File: app/src/main/java/com/example/mobile6/ui/medicine/MedicineSearchFragment.kt
package com.example.mobile6.ui.medicine

import android.os.Bundle
import android.view.LayoutInflater
import android.view.ViewGroup
import androidx.core.widget.doOnTextChanged
import androidx.fragment.app.viewModels
import androidx.lifecycle.Lifecycle
import androidx.lifecycle.lifecycleScope
import androidx.lifecycle.repeatOnLifecycle
import androidx.navigation.NavOptions
import androidx.recyclerview.widget.LinearLayoutManager
import com.example.mobile6.R
import com.example.mobile6.databinding.FragmentMedicineSearchBinding
import com.example.mobile6.ui.adapter.MedicineAdapter
import com.example.mobile6.ui.base.BaseFragment
import com.example.mobile6.ui.util.defaultAnim
import dagger.hilt.android.AndroidEntryPoint
import kotlinx.coroutines.flow.collectLatest
import kotlinx.coroutines.launch

@AndroidEntryPoint
class MedicineSearchFragment : BaseFragment<FragmentMedicineSearchBinding>() {

    private val viewModel: MedicineListViewModel by viewModels()
    private lateinit var medicineAdapter: MedicineAdapter

    override val bindingInflater: (LayoutInflater, ViewGroup?) -> FragmentMedicineSearchBinding
        get() = { inflater, container ->
            FragmentMedicineSearchBinding.inflate(inflater, container, false)
        }

    override fun initViews() {
        binding.backButton.setOnClickListener {
            back()
        }

        setupRecyclerView()
        setupSearchView()
    }

    private fun setupRecyclerView() {
        medicineAdapter = MedicineAdapter { medicine ->
            val bundle = Bundle().apply {
                putLong("medicineId", medicine.id);
            }
            navigateTo(
                R.id.action_medicineSearchFragment_to_medicineDetailFragment,
                bundle
            )
        }

        binding.medicinesRecyclerView.apply {
            adapter = medicineAdapter
            layoutManager = LinearLayoutManager(requireContext())
        }
    }

    private fun setupSearchView() {
        binding.searchEditText.doOnTextChanged { text, _, _, _ ->
            viewModel.searchMedicines(text.toString())
        }
    }

    override fun initObservers() {
        viewLifecycleOwner.lifecycleScope.launch {
            viewLifecycleOwner.repeatOnLifecycle(Lifecycle.State.STARTED) {
                viewModel.uiState.collectLatest { state ->
                    if (state.isLoading) {
                        handleLoadingState(true)
                        return@collectLatest
                    }
                    if (state.error != null) {
                        handleErrorState(state.error)
                        return@collectLatest
                    }
                    medicineAdapter.submitList(state.medicines)
                }
            }
        }
    }

    private fun handleLoadingState(isLoading: Boolean) {
        // Show/hide loading indicator if you have one
        // binding.progressBar.isVisible = isLoading
    }

    private fun handleErrorState(error: String?) {
        // Show error message if needed
        // error?.let {
        //     Toast.makeText(requireContext(), it, Toast.LENGTH_SHORT).show()
        // }
    }
}